### PR TITLE
Replace all `Lock` to `RLock`(Fixes #7686)

### LINKED ISSUE #7686: Deadlock
# Disclaimer
I don't know much about celery yet.
So, I create a Issue, not a PR, to report this.

I seen Celery is uses lot `theating.Lock()`, and multiple calls `acquire()`(same as `with Lock():`) in same thread.
This is the point of deadlock problem.

I have sent a [PR](https://github.com/celery/celery/pull/7652) to similar problem, it was merged to master.
It looks work normal.
But actually, I'm not completely sure is this PR stable or not.
I'm not sure if replacing most `Lock()` with `RLock()` in Celery, will cause other bugs or not.

# Bug description
I don't know what wrong will be cause with this deadlock.
But it must be a deadlock.

In there, It got the Lock().
https://github.com/celery/celery/blob/bdbf6d6ae1aca9addd81800b5dd2e8c3477afb18/celery/events/state.py#L444
Then, It will got Lock() again.
https://github.com/celery/celery/blob/bdbf6d6ae1aca9addd81800b5dd2e8c3477afb18/celery/events/state.py#L452

### PR DESCRIPTION
**Pull request by 954-Ivory**
## Description
Try to fixed most of deadlock.
Replace all `Lock` to `RLock`(Fixes #7686).

### CODE REVIEW

**Comment by 954-Ivory** [comment]
I get the failing reason from CI log.
But I have some question:
1. How should I do now? Close this PR and send a new PR?
2. Can I Fork CI in my Repo? I want to checking again when i changed code.

**Comment by 954-Ivory** [comment]
I use pycharm to `Edit Commit Message`, the commit seens a little dirty.
But this PR seens work well now.

**Comment by auvipy** [comment]
no issue, we will squash commit.

File: celery/events/dispatcher.py [hunk 1]
```diff
@@ -218,7 +218,10 @@ def extend_buffer(self, other):
 
     def close(self):
         """Close the event dispatcher."""
-        self.mutex.locked() and self.mutex.release()
+        try:
+            self.mutex.release()
+        except RuntimeError:
+            pass
```
Comment (thedrow) [inline celery/events/dispatcher.py:224]: This is too broad. We need to catch specific exceptions.

**Review by thedrow — CHANGES_REQUESTED** [review]

File: celery/events/dispatcher.py (hunk 1 above)
Comment (954-Ivory) [reply celery/events/dispatcher.py:224]: This just catch the `RuntimeError` from `RLock.release()`.
But I think maybe CI exception cause by RLock didn't release.

**Comment by thedrow** [comment]
I'm a little bit concerned that we're replacing all locks with reentrant locks without any evidence it will resolve the deadlock (although it makes some sense).
Perhaps you can use [pydld](https://github.com/mindstorm38/pydld) or a similar tool to verify this.

**Comment by auvipy** [comment]
@954-Ivory can you revisit this please?

**Comment by 954-Ivory** [comment]
> @954-Ivory can you revisit this please?

**Maybe we should consider temporarily closing this PR.**

I checked the document about `Lock` and `RLock`, there has a little different:

#### `Lock.release()`
_This can be called from any thread, not only the thread which has acquired the lock._
https://docs.python.org/3/library/threading.html?highlight=rlock#threading.Lock.release

#### `RLock.release()`
_Only call this method when the calling thread owns the lock._
https://docs.python.org/3/library/threading.html?highlight=rlock#threading.RLock.release

**I'm not sure about that, does celery has any scene, about release a thread lock in the not lock owning thread.**

**Pull request closed without merging**